backend/server/routers/courses.py

```diff
@@ -129,23 +135,45 @@ def search(string):
           “COMP1531”: “SoftEng Fundamentals, 
             ……. }
     """
-    # TODO: is regex search really something we want?
-    # malicious regex can cause DOS depending on regex implementation
-    # Would fuzzy search be better?
-    pat = re.compile(r"{}".format(string), re.I)
-    code_query = list(coursesCOL.find({"code": {"$regex": pat}}))
-    title_query = list(coursesCOL.find({"title": {"$regex": pat}}))
-
-    if not code_query and not title_query:
-        for year in sorted(ARCHIVED_YEARS, reverse=True):
-            code_query = list(archivesDB[str(year)].find({"code": {"$regex": pat}}))
-            title_query = list(archivesDB[str(year)].find({"title": {"$regex": pat}}))
-            if code_query or title_query:
-                break
+    global ALL_COURSES
+    from server.routers.programs import getStructure
 
-    return {
-        course["code"]: course["title"] for course in chain(code_query, title_query)
-    }
+    if ALL_COURSES is None:
+        ALL_COURSES = [{"title": course["title"], "code": course["code"]} for course in coursesCOL.find()]
+
+    # TODO: consider ways to search legacy courses
+    # TODO: can you have a minor without a major selected?
+    #       will wreak havoc on the argument order with *specialisations
+    specialisations = list(userData.specialisations.keys())
+    structure = getStructure(userData.program, *specialisations)['structure']
+
+    top_results = sorted(ALL_COURSES, reverse=True,
```

Comment:
so we will probably need to add legacy courses back. to do this, you should chain the iterables together, and give a low/minimal weighting to those legacy courses.

Reply:
The whole legacy courses structure is going to need an overhaul for this. It worked for the previous implementation since you'd only check legacy years if the course didn't exist in the current year, but since we're fuzzy matching and will have to add all the previous years courses this gets problematic due to all of the duplicates from courses that have been in the handbook for multiple years.

Reply:
A side effect of all the duplicates is that the search call is noticeably slower when running the weighting algos over every course from every year

Reply:
Unless you have a straightforward solution, this might be better left as a separate issue.

Reply:
hmm would it be simple to do a `fetch_latest_offering` for all courses? So basically you look through the archive, and if the course is not currently in the list of all courses, its added - with the list traversed backwards. This might be expensive, but its probably only going to need to be done once. This helps get rid of duplicates.

Reply:
I'll look into it. While I'm doing this, do we want to indicate to the user that the course is a legacy course somehow?

Reply:
@joooooooooooooooooooooooooooooooooooosh I think we do this in 2 ways - we have a note in the course selector and another in degree planner (which is a warning). Should we add an icon in search as well?

Reply:
potentially, i'll think about it